Refactor: PureScript

gaku-sei opened the pull request:
The current parser and logic is error prone and inconsistent, we're still having errors when generating some code in PureScript (the `Over`/`\/` classes get removed). The very first version was a very simple one shot to generate some very simple Elm classes, but Tailwind got more complex over time, and so did our "parser".

PureScript allows us not only to have some better type safety, but also to increase the consistency, and the readability. It also comes with an excellent cli library: https://pursuit.purescript.org/packages/purescript-optparse/3.0.0

- [x] Overall parser for css classes (event the most "complex" ones, like spacing)
- [x] Templating system (using handlebars)
- [x] PureScript template and logic implemented (and working)
- [x] Tests for the parser
- [x] CLI (the api should be 1:1 with the current one, no breaking changes expected)
- [x] Handle all the CLI options
- [x] Use the tailwind build function
- [x] Elm...
- [x] ... TypeScript ...
- [x] ... and ReasonML templates
- [x] Add a TypeScript type level generator that mimics https://github.com/vadimdemedes/tailwind-rn api
- [x] Test more and polish the code

gaku-sei commented on src/Main.purs:
```diff
@@ -0,0 +1,23 @@
+  app = do
+    { config, cssInput, cssOutput, output } <- ask
+    tailwindBuild config cssInput cssOutput
+    catchAndKill $ mkdirp output
```
`catchAndKill` is an util function that catches the exceptions, and terminates (kills) the program (exit code 1)

gaku-sei reviewed (commented):
I also arbitrary chose to have the util and parser functions "outside" of the application (not using the AppM monad), while main and printer are "inside" the application, and make use of AppM.

The rational is mostly conceptual, parsers and utils are not bounded to an app.

I don't mind changing this if needed 😄

dandoh commented on src/Utils.js:
```diff
@@ -0,0 +1,53 @@
+                return acc;
+              }
+
+              return [...acc, className.replace(/\\/, "\\\\")];
```
Why do we need this?

dandoh commented on src/Parser.purs:
```diff
@@ -0,0 +1,64 @@
+      , string "\\:"
+      , string "-"
+      , string "\\/"
+      ]
```
How is it different if we just return this in Js:
```
return [...acc, className];
```
And here it will be:
```
  anyValidAndSpecialChar =
    choice
      [ const "-over-" <$> string "\\/"
      , const "-" <$> string "\\:"
      , singleton <$> alphaNum
      ]
``` 
I'm pretty sure I'm missing something, but just out of curiousity xD

gaku-sei replied:
Unfortunately we need to escape \ at one point, otherwise when printed by Handlebars we'll lose the \ and the string will sometimes be invalid ("\:" is not a valid string while "\\:" is).
So at one point I had to add some \ logic 😞 
That being said, that logic could be moved in PureScript, we just need a ClassNameParse that would replace the occurrences lf \ by \\

Also, i looked for an option in Handlebars but couldn'tt fond any

gaku-sei replied:
But for the last 3 parsers it's just a copy paste mistake xD
Let me removed it ^^
(I'd like to remove the references to AppM they're not really useful and quite confusing 😕)

gaku-sei commented:
@dandoh I cleaned up the parsers (and the AppM references), now the valid characters parser is defined as follow:

```purescript
  anyValidAndSpecialChar =
    choice
      [ const "-over-" <$> string "\\\\/"
      , const "-" <$> string "\\\\:"
      , singleton <$> alphaNum
      , string "-"
      ]
```

(Notice that, as said above and as you noticed, we may be able to get rid off that part from the js code, just need to escape the class name before we print it using Handlebars).

gaku-sei commented:
I also fixed the binary (shebang + wrong paths), and fixed an edge case where the class name has a `-` modifier after a `:` character: `hover\\\\:-translate-x-1\\\\/2` should be transformed to `hoverNegTranslateX1Over2` but used to be transformed to `hoverTranslateX1Over2`.

gaku-sei commented:
Tested on https://github.com/scoville/experimental-schools-manager and it works (it actually fixed some formerly wrongly generated names, like `text2xl` instead of `text2Xl`)

gaku-sei commented:
Thank @dandoh 

There are still a couple of tasks we can work on before release:
- Testing on encourage client and admin
- Splitting the Printer module as adviced here: https://github.com/scoville/tailwind-generator/pull/22#discussion_r502928140
- Merging the new verbosity mode https://github.com/scoville/tailwind-generator/pull/22 😄
